File: ggz/src/ggzcore/player.c
/*
 * File: player.c
 * Author: Brent Hendricks
 * Project: GGZ Core Client Lib
 * Date: 6/5/00
 * $Id: player.c 6785 2005-01-21 18:48:01Z jdorje $
 *
 * This fils contains functions for handling players
 *
 * Copyright (C) 1998 Brent Hendricks.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>		/* Site-specific config */
#endif

#include <stdlib.h>
#include <string.h>

#include <ggz.h>

#include "player.h"
#include "room.h"

/* Publicly exported functions */

char* ggzcore_player_get_name(GGZPlayer *player)
{
	if (!player)
		return NULL;

	return _ggzcore_player_get_name(player);
}


GGZPlayerType ggzcore_player_get_type(GGZPlayer *player)
{
	if (!player)
		return -1;

	return _ggzcore_player_get_type(player);
}


GGZTable* ggzcore_player_get_table(GGZPlayer *player)
{
	if (!player)
		return NULL;

	return _ggzcore_player_get_table(player);
}


int ggzcore_player_get_lag(GGZPlayer *player)
{
	if (!player)
		return 0;

	return _ggzcore_player_get_lag(player);
}



int ggzcore_player_get_record(GGZPlayer *player,
			      int *wins, int *losses,
			      int *ties, int *forfeits)
{
	if (!player || !wins || !losses || !ties || !forfeits)
		return 0;
	return _ggzcore_player_get_record(player, wins, losses,
					  ties, forfeits);
}

int ggzcore_player_get_rating(GGZPlayer *player, int *rating)
{
	if (!player || !rating)
		return 0;
	return _ggzcore_player_get_rating(player, rating);
}

int ggzcore_player_get_ranking(GGZPlayer *player, int *ranking)
{
	if (!player || !ranking)
		return 0;
	return _ggzcore_player_get_ranking(player, ranking);
}

int ggzcore_player_get_highscore(GGZPlayer *player, int *highscore)
{
	if (!player || !highscore)
		return 0;
	return _ggzcore_player_get_highscore(player, highscore);
}


/* 
 * Internal library functions (prototypes in player.h) 
 * NOTE:All of these functions assume valid inputs!
 */

struct _GGZPlayer* _ggzcore_player_new(void)
{
	struct _GGZPlayer *player;

	player = ggz_malloc(sizeof(struct _GGZPlayer));

	/* Set to invalid table */
	player->table = -1;

	/* Assume no lag */
	player->lag = -1;

	player->wins = NO_RECORD;
	player->losses = NO_RECORD;
	player->ties = NO_RECORD;
	player->forfeits = NO_RECORD;
	player->rating = NO_RATING;
	player->ranking = NO_RANKING;
	player->highscore = NO_HIGHSCORE;

	return player;
}


void _ggzcore_player_init(struct _GGZPlayer *player, 
			  const char *name, 
			  struct _GGZRoom *room,
			  const int table,
			  const GGZPlayerType type,
			  const int lag)
{
	player->name = ggz_strdup(name);
	player->room = room;
	player->table = table;
	player->type = type;
	player->lag = lag;
}


void _ggzcore_player_init_stats(GGZPlayer *player,
				const int wins,
				const int losses,
				const int ties,
				const int forfeits,
				const int rating,
				const int ranking,
				const long highscore)
{
	player->wins = wins;
	player->losses = losses;
	player->ties = ties;
	player->forfeits = forfeits;
	player->rating = rating;
	player->ranking = ranking;
	player->highscore = highscore;
}


void _ggzcore_player_free(struct _GGZPlayer *player)
{
	if (player->name)
		ggz_free(player->name);

	ggz_free(player);
}


void _ggzcore_player_set_table(struct _GGZPlayer *player, const int table)
{
	player->table = table;
}


void _ggzcore_player_set_lag(struct _GGZPlayer *player, const int lag)
{
	player->lag = lag;
}


char* _ggzcore_player_get_name(struct _GGZPlayer *player)
{
	return player->name;
}


GGZPlayerType _ggzcore_player_get_type(struct _GGZPlayer *player)
{
	return player->type;
}


GGZTable* _ggzcore_player_get_table(struct _GGZPlayer *player)
{
	if (player->table == -1)
		return NULL;
	
	return ggzcore_room_get_table_by_id(player->room, player->table);
}


int _ggzcore_player_get_lag(struct _GGZPlayer *player)
{
	return player->lag;
}


int _ggzcore_player_get_record(GGZPlayer *player,
				int *wins, int *losses,
				int *ties, int *forfeits)
{
	if (player->wins == NO_RECORD
	    && player->losses == NO_RECORD
	    && player->ties == NO_RECORD
	    && player->forfeits == NO_RECORD)
		return 0;

#ifndef MAX
#  define MAX(a, b) ((a) > (b) ? (a) : (b))
#endif

	/* NO_RECORD is -1.  If we have a stat for anything, we should
	   return all stats and assume 0 for any we don't know. */

	*wins = MAX(player->wins, 0);
	*losses = MAX(player->losses, 0);
	*ties = MAX(player->ties, 0);
	*forfeits = MAX(player->forfeits, 0);

	return 1;
}


int _ggzcore_player_get_rating(GGZPlayer *player,
				int *rating)
{
	if (player->rating == NO_RATING)
		return 0;
	*rating = player->rating;
	return 1;
}


int _ggzcore_player_get_ranking(GGZPlayer *player,
				int *ranking)
{
	if (player->ranking == NO_RANKING)
		return 0;
	*ranking = player->ranking;
	return 1;
}


int _ggzcore_player_get_highscore(GGZPlayer *player, int *highscore)
{
	if (player->highscore == NO_HIGHSCORE)
		return 0;
	*highscore = player->highscore;
	return 1;
}


int _ggzcore_player_compare(const void *p, const void *q)
{
	const GGZPlayer *s_p = p;
	const GGZPlayer *s_q = q;

	return strcmp(s_p->name, s_q->name);
}


void* _ggzcore_player_create(void* p)
{
	struct _GGZPlayer *new, *src = p;

	new = _ggzcore_player_new();
	_ggzcore_player_init(new, src->name, src->room, src->table, src->type, src->lag);

	return (void*)new;
}


void  _ggzcore_player_destroy(void* p)
{
	_ggzcore_player_free(p);
}

/* Static functions internal to this file */

File: java/src/ggz/client/core/Player.java
/*
 * Copyright (C) 2006  Helg Bredow
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ggz.client.core;

import ggz.common.Perm;
import ggz.common.PermSet;
import ggz.common.PlayerType;

import java.io.IOException;
import java.text.Collator;
import java.util.Comparator;

/*
 * The Player structure is meant to be a node in a list of the players in the
 * current room .
 */
public class Player {
    public static final int NO_RECORD = -1;

    public static final int NO_RATING = 0;

    public static final int NO_RANKING = 0;

    public static final int NO_HIGHSCORE = -1;

    /* Name of player */
    private String name;

    /* Type of player */
    private PlayerType type;

    private PermSet perms;

    /* Pointer to room player is in */
    private Room room;

    /* Server ID of table player is at */
    private int table;

    /* Lag of the player */
    private int lag;

    /* Record of the player (or NO_RECORD). */
    public int wins, losses, ties, forfeits;

    /* Rating of the player (or NO_RATING) */
    private int rating;

    /* Ranking of the player (or NO_RANKING) */
    private int ranking;

    /* Player's highest score (or NO_HIGHSCORE) */
    private int highscore;

    public String get_name() {
        return this.name;
    }

    public int get_rating() {
        return this.rating;
    }

    public int get_ranking() {
        return this.ranking;
    }

    public int get_highscore() {
        return this.highscore;
    }

    public int get_wins() {
        return this.wins;
    }

    public int get_losses() {
        return this.losses;
    }

    public int get_ties() {
        return this.ties;
    }

    public int get_forfeits() {
        return this.forfeits;
    }

    /**
     * Constructor used only by RoomChatPanel to calculate column widths.
     * 
     * @param name
     */
    public Player(String name) {
        this.name = name;
    }

    /*
     * Internal library functions (prototypes in player.h) NOTE:All of these
     * functions assume valid inputs!
     */

    Player(String name, Room room, int table, PlayerType type, PermSet perms,
            int lag) {
        this.wins = NO_RECORD;
        this.losses = NO_RECORD;
        this.forfeits = NO_RECORD;
        this.ties = NO_RECORD;
        this.rating = NO_RATING;
        this.ranking = NO_RANKING;
        this.highscore = NO_HIGHSCORE;
        this.name = name;
        this.room = room;
        this.table = table;
        this.type = type;
        this.perms = perms;
        this.lag = lag;
    }

    void set_stats(int new_wins, int new_losses, int new_ties,
            int new_forfeits, int new_rating, int new_ranking, int new_highscore) {
        this.wins = new_wins;
        this.losses = new_losses;
        this.ties = new_ties;
        this.forfeits = new_forfeits;
        this.rating = new_rating;
        this.ranking = new_ranking;
        this.highscore = new_highscore;
    }

    void set_table(int table) {
        this.table = table;
    }

    void set_lag(int lag) {
        this.lag = lag;
    }

    public PlayerType get_type() {
        return this.type;
    }

    public PermSet get_perms() {
        return this.perms;
    }

    public Room get_room() {
        return this.room;
    }

    public Table get_table() {
        if (this.table == -1)
            return null;

        return this.room.get_table_by_id(this.table);
    }

    public int get_lag() {
        return this.lag;
    }

    public boolean has_perm(Perm perm) {
        return this.perms != null && this.perms.isSet(perm);
    }

    /**
     * Sends a request to the server to set the permission. The permission is
     * not set in this player object directly.
     * 
     * @param perm
     * @param set
     * @throws IOException
     */
    public void set_perm(Perm perm, boolean set) throws IOException {
        Server server = get_room().get_server();
        server.set_perm(get_name(), perm, set);
    }

    /**
     * Sens a series of requests to the server to set the given permissions.
     * 
     * @param perms
     * @throws IOException
     */
    protected void set_perms(PermSet perms) {
        this.perms = perms;
    }

    protected void set_type(PlayerType type) {
        this.type = type;
    }

    public boolean equals(Object o) {
        // Player names are not case sensitive.
        return (o != null) && (o instanceof Player)
                && this.name.equalsIgnoreCase(((Player) o).name);
    }

    public String toString() {
        return get_name();
    }

    /**
     * This is here only because the Java API says we should implement
     * hashCode() if we implement equals().
     */
    public int hashCode() {
        return this.name.hashCode();
    }

    public static final Comparator SORT_BY_NAME = new Comparator() {
        private Collator collator = Collator.getInstance();

        public int compare(Object o1, Object o2) {
            Player p1 = (Player) o1;
            Player p2 = (Player) o2;
            if (p1 == null && p2 == null) {
                return 0;
            } else if (p1 == null) {
                return -1;
            } else if (p2 == null) {
                return 1;
            } else {
                return collator.compare(p1.get_name(), p2.get_name());
            }
        }
    };

    public static final Comparator SORT_BY_TYPE = new Comparator() {

        public int compare(Object o1, Object o2) {
            Player p1 = (Player) o1;
            Player p2 = (Player) o2;
            if (p1 == null && p2 == null) {
                return 0;
            } else if (p1 == null) {
                return -1;
            } else if (p2 == null) {
                return 1;
            } else {
                PlayerType type1 = p1.get_type();
                PlayerType type2 = p2.get_type();
                int result = PlayerType.SORT_BY_PRIVILEGE.compare(type1, type2);
                if (result == 0) {
                    return SORT_BY_NAME.compare(p1, p2);
                }
                return result;
            }
        }
    };

}
